### DISCUSSION TITLE
How to support file uploads in webView and opening external links in the default browser

### DISCUSSION THREAD

**Discussion by ukane-philemon** [progrium/darwinkit#143] [General]
Hi, @progrium. I appreciate the Macdrive library and have been trying to use it to achieve certain functionality in Go.

- Supporting file uploads: 
     I have implemented the `webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:` method and have set the webView UIDelegate but I get a panic each time it gets to `panel.Send("beginWithCompletionHandler:", completionHandler)` :
     ```
     webkitUIDelegate := objc.NewClass("WKUIDelegate", "NSObject").Alloc().Init()
	webkitUIDelegate.Class().AddMethod("webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:", func(_ objc.Object, webView objc.Object, openPanelParams objc.Object, frame objc.Object, completionHandler objc.Object) {
		allowMultipleSelection := openPanelParams.Get("allowsMultipleSelection")
		allowDirectories := openPanelParams.Get("allowsDirectories")
		panel := objc.Get("NSOpenPanel").Send("openPanel")
		panel.Set("canChooseFiles:", core.True)
		panel.Set("canChooseDirectories:", allowDirectories)
		panel.Set("allowsMultipleSelection:", allowMultipleSelection)
		panel.Send("beginWithCompletionHandler:", completionHandler)
     })
     ```
-  Opening external links in the default browser:
         I have also implemented the `webView:decidePolicyForNavigationAction:decisionHandler:` method but was unable to execute the decision handler in Go using Macdrive.
         
    These two issues have been a pita for some time so I thought about asking for help. Thanks for your reply in advance.
    
    PS: I'd appreciate it if you can also show me a snippet of how to execute a completion handler(with/without arguments) using the Macdrive library in Go.
    PSS: I used a workaround for this. You can see my implementation here: https://gist.github.com/ukane-philemon/161467e65b91eaade6561a395cc1d568

**Comment by progrium** [comment]
It's on our list to figure out how to handle callbacks. If you're adventurous, you can use CGO to write Obj-C to work around this in an ad-hoc fashion.

**Reply by ukane-philemon** [reply]
Oh. Tbh, I've spent several days trying to figure out how to handle the callbacks but will look into how I can work around this until then.  Thanks for your suggestion.

**Reply by ukane-philemon** [reply]
I wrote some Obj-C code to handle the two issues mentioned in my [original comment](https://github.com/progrium/macdriver/discussions/143#discussion-5399530) (I had to learn and play around with Obj-C but it was fun).

I could contribute an example/code for these scenarios. Let me know how you'd want me to go about it, @progrium.

**Reply by progrium** [reply]
You could share it as a gist here for now.

**Reply by fdidron** [reply]
@ukane-philemon I'd love to see how you implemented callbacks :)

**Reply by ukane-philemon** [reply]
Here: https://gist.github.com/ukane-philemon/161467e65b91eaade6561a395cc1d568

**Reply by fdidron** [reply]
Thank you

**Comment by progrium** [comment]
Just documenting that there should be a delegate builder and callback support for all this to work now with the new version in main.
